File: go/replay.go
package l402kit

import (
	"crypto/sha256"
	"encoding/hex"
	"sync"
)

// replayStore is an in-memory set of seen preimage hashes.
// For multi-instance deployments, replace with Redis or a shared DB.
var replayStore = struct {
	mu   sync.Mutex
	seen map[string]struct{}
}{seen: make(map[string]struct{})}

// CheckAndMarkPreimage returns true if the preimage has NOT been seen before,
// and atomically marks it as used. Returns false if it was already used.
func CheckAndMarkPreimage(preimage string) bool {
	h := sha256.Sum256([]byte(preimage))
	key := hex.EncodeToString(h[:])

	replayStore.mu.Lock()
	defer replayStore.mu.Unlock()

	if _, exists := replayStore.seen[key]; exists {
		return false
	}
	replayStore.seen[key] = struct{}{}
	return true
}

File: rust/src/replay.rs
use once_cell::sync::Lazy;
use sha2::{Digest, Sha256};
use std::collections::HashSet;
use std::sync::Mutex;

// In-memory anti-replay store.
// For multi-instance deployments, replace with Redis or a shared database.
static REPLAY_STORE: Lazy<Mutex<HashSet<String>>> =
    Lazy::new(|| Mutex::new(HashSet::new()));

/// Returns `true` if the preimage has NOT been seen before, and atomically
/// marks it as used. Returns `false` if the preimage was already used (replay attack).
pub fn check_and_mark_preimage(preimage: &str) -> bool {
    let mut hasher = Sha256::new();
    hasher.update(preimage.as_bytes());
    let key = hex::encode(hasher.finalize());

    let mut store = REPLAY_STORE.lock().unwrap_or_else(|e| e.into_inner());

    if store.contains(&key) {
        return false;
    }
    store.insert(key);
    true
}
